zebra-consensus/src/checkpoint.rs

```diff
@@ -384,36 +375,47 @@ impl CheckpointVerifier {
     ///
     /// If the block does not have a coinbase height, sends an error on `tx`,
     /// and does not queue the block.
-    fn queue_block(&mut self, block: Arc<Block>) -> oneshot::Receiver<Result<block::Hash, Error>> {
+    pub(super) fn queue_block(
+        &mut self,
+        block: Arc<Block>,
+    ) -> oneshot::Receiver<Result<block::Hash, Error>> {
         // Set up a oneshot channel to send results
         let (tx, rx) = oneshot::channel();
 
         // Check for a valid height
-        let height = match self.check_block(&block) {
-            Ok(height) => height,
-            Err(error) => {
-                // Block errors happen frequently on mainnet, due to bad peers.
+        let height = match block.coinbase_height() {
+            Some(height) => height,
+            None => {
+                let error = "the block does not have a coinbase height".into();
                 tracing::debug!(?error);
-
-                // Sending might fail, depending on what the caller does with rx,
-                // but there's nothing we can do about it.
                 let _ = tx.send(Err(error));
                 return rx;
             }
         };
 
-        // Since we're using Arc<Block>, each entry is a single pointer to the
-        // Arc. But there are a lot of QueuedBlockLists in the queue, so we keep
+        let hash = block.hash();
+
+        // Treat duplicate verified blocks as successful verifies.
+        // Otherwise, the syncer restarts twice on errors, once on the
+        // original error, and again as it re-submits duplicate blocks.
+        if let Err(err) = self.check_height(height) {
+            // Either the block is behind the initial tip from the state, or it
+            // has already been verified by a previous checkpoint.
+            tracing::info!(height = ?height, hash = ?hash, err = ?err);
+            let _ = tx.send(Ok(hash));
+            return rx;
+        }
+
+        // There are a lot of QueuedBlockLists in the queue, so we keep
         // allocations as small as possible.
         let qblocks = self
             .queued
             .entry(height)
             .or_insert_with(|| QueuedBlockList::with_capacity(1));
 
-        let hash = block.hash();
-
         for qb in qblocks.iter_mut() {
             if qb.hash == hash {
+                // TODO: don't restart the syncer on duplicate blocks
                 let old_tx = std::mem::replace(&mut qb.tx, tx);
                 let e = "rejected older of duplicate verification requests".into();
                 tracing::debug!(?e);
```

Comment:
If the syncer is operating correctly, and we have good peers, then this code should be unreachable, because:
* verified blocks are added to the state
* the block locator starts at the state tip
* peers should extend the block locator tip forwards

Also, when the syncer restarts, it drops the old channels, so it should never get these errors. But I've seen them, probably due to a race between the sync restart, and the async download and verify tasks.

If we have bad peers, then we want to restart. If we have peers that aren't synced yet, then we might want to restart to pick up missing blocks.

So I don't think this change is particularly important.